compiler/src/iree/compiler/Codegen/Common/MaterializeTuningSpecsPass.cpp

```diff
@@ -0,0 +1,160 @@
+    llvm::cl::desc("File path to a module containing a tuning spec (transform "
+                   "dialect library)."),
+    llvm::cl::init(""));
+
+llvm::cl::opt<std::string> clCodegenTuningSpecDumpDir(
+    "iree-codegen-dump-tuning-specs-to",
+    llvm::cl::desc(
+        "Dump the final tuning spec modules to the specified directory. When "
+        "set to '-', prints the tuning spec to stdout."),
+    llvm::cl::init(""));
+
+using mlir::transform::NamedSequenceOp;
+
+static LogicalResult dumpFinalTuningSpecToDir(ModuleOp tuningSpec,
+                                              StringRef dir) {
+  if (dir == "-") {
+    tuningSpec->print(llvm::outs());
+    return success();
+  }
+
+  llvm::sys::fs::create_directories(dir);
+  llvm::SmallString<64> dumpPath;
+  auto dumpFileEC = llvm::sys::fs::createUniqueFile(
+      Twine(dir) + "/iree_tuning_spec_%%.mlir", dumpPath);
+  if (dumpFileEC) {
+    return tuningSpec->emitError()
+           << "Failed to create a unique file in " << dir << "\n";
+  }
+  LDBG("Linked tuning spec file path: " << dumpPath);
+
+  std::string error;
+  auto file = mlir::openOutputFile(dumpPath, &error);
+  if (!file) {
+    return tuningSpec->emitError()
+           << "Failed to open a tuning spec dump file " << dumpPath << "\n";
+  }
+
+  tuningSpec->print(file->os());
+  file->keep();
+  return success();
+}
+
+static FailureOr<DenseElementsAttr>
+serializeTuningSpecToAttr(ModuleOp tuningSpec) {
+  std::string buffer;
+  llvm::raw_string_ostream os(buffer);
+  if (failed(writeBytecodeToFile(tuningSpec, os))) {
+    return failure();
+  }
+
+  auto bufferSize = static_cast<int64_t>(buffer.size());
+  auto bufferShape = VectorType::get(
+      bufferSize, IntegerType::get(tuningSpec->getContext(), 8));
+  return DenseElementsAttr::getFromRawBuffer(
+      bufferShape, ArrayRef(buffer.data(), buffer.data() + bufferSize));
+}
+
+struct MaterializeTuningSpecsPass final
+    : impl::MaterializeTuningSpecsPassBase<MaterializeTuningSpecsPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registerTransformDialectTranslationDependentDialects(registry);
+  }
+
+  void runOnOperation() override {
+    if (clCodegenTuningSpecPath.empty()) {
+      return;
+    }
+
+    ModuleOp module = getOperation();
+    MLIRContext *ctx = &getContext();
+    auto dialect = ctx->getOrLoadDialect<IREE::Codegen::IREECodegenDialect>();
+    auto maybeTransformLibrary =
+        dialect->getOrLoadTransformLibraryModule(clCodegenTuningSpecPath);
+    if (failed(maybeTransformLibrary)) {
+      module->emitError()
+          << "Failed to load tuning spec transform dialect library from "
+          << clCodegenTuningSpecPath;
+      return signalPassFailure();
+    }
+
+    ModuleOp userTuningSpec = *maybeTransformLibrary;
+    if (!userTuningSpec.getSymName()) {
+      // Set a module name so that we can refer to its nested symbols.
+      userTuningSpec.setSymName("iree_user_tuning_spec");
```

Comment:
nit: move this out as `constexpr char* kUserTuningSpaceName`

Reply:
This is just a name, I don't expect any other code to rely on it.